Adds Users Swagger APIKeyAuthenticator

nmagnezi opened the pull request:
TODO:

* (Later) test-infra
Q: Error codes --> we currently return 501 for exp token

 Co-authored-by: Fred Rolland frolland@redhat.com

nmagnezi commented on cmd/main.go:
```diff
@@ -139,6 +141,12 @@ func main() {
 	events := events.NewApi(eventsHandler, logrus.WithField("pkg", "eventsApi"))
 
+
+	var authMiddleware auth.JWTMiddleware = &auth.AuthMiddlewareMock{}
```
Per request from @ronniel1 , this should go into the middleware and not be handled by main.

filanov replied:
+1

ronniel1 commented on Makefile:
```diff
@@ -28,7 +28,8 @@ lint:
 .PHONY: build
-build: create-build-dir lint unit-test
+## TODO(nmagnezi): Removed lint unit-test for now.
+build: create-build-dir
```
Make sure not to commit this

ronniel1 commented on cmd/main.go:
```diff
@@ -139,6 +141,12 @@ func main() {
 	events := events.NewApi(eventsHandler, logrus.WithField("pkg", "eventsApi"))
 
+
+	var authMiddleware auth.JWTMiddleware = &auth.AuthMiddlewareMock{}
```
why are we initializing with a mock?

avishayt commented:
@nmagnezi Did we solve the issue of turning off authentication for disconnected on-prem (including client)?  How does it work in test env?  Thanks!

eranco74 commented on cmd/main.go:
```diff
@@ -147,10 +155,11 @@ func main() {
 		ManagedDomainsAPI: domainHandler,
 		InnerMiddleware:   metrics.WithMatchedRoute(log.WithField("pkg", "matched-h")),
 	})
+	h = authMiddleware.AuthenticateAccountJWT(h)
```
in case Options.Auth.EnableJWT option is false this will fail

nmagnezi replied:
Done

nmagnezi commented on pkg/app/middleware.go:
```diff
@@ -28,3 +31,17 @@ func WithHealthMiddleware(next http.Handler) http.Handler {
 }
+
+// WithAuthMiddleware returns authentication middleware that serves both users and agents
+func WithAuthMiddleware(next http.Handler) http.Handler {
```
should main pass authConfig as param to WithAuthMiddleware() ?

ronniel1 replied:
yes it should

However, the middleware needs to be an inner-middleware

ronniel1 commented on pkg/auth/auth_middleware.go:
```diff
@@ -0,0 +1,141 @@
+func (a *AuthMiddleware) AuthenticateAccount(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		route := middleware.MatchedRouteFrom(r)
```
This will work only on an inner-middleware

ronniel1 commented on pkg/logger/logger.go:
```diff
@@ -0,0 +1,112 @@
+	"fmt"
+
+	"github.com/getsentry/sentry-go"
+	"github.com/golang/glog"
```
We are using logrus, not glog

nmagnezi replied:
I'm going to remove this and use the already existing pkg/log instead

ronniel1 reviewed (commented):
I don't see UTs for most of the stuff

ronniel1 commented on pkg/auth/helpers.go:
```diff
@@ -0,0 +1,133 @@
+	"net/http"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/golang/glog"
```
use logrus instead of glog

nmagnezi replied:
Done

ronniel1 commented on cmd/main.go:
```diff
@@ -158,18 +160,21 @@ func main() {
 	} else {
 		log.Info("Disabled image expiration monitor")
 	}
+	auth.InitAuthHandler(Options.Auth.JwkCertURL)
```
suggest to change auth to be a class that is initialized with 'New()'.
As it is now, we initialize a global auth

New() should accept the options and a logger

ronniel1 commented on cmd/main.go:
```diff
@@ -158,18 +160,21 @@ func main() {
-		InnerMiddleware:   metrics.WithMatchedRoute(log.WithField("pkg", "matched-h")),
+		AuthAgentAuth:       auth.AuthAgentAuth,
+		AuthUserAuth:        auth.AuthUserAuth,
+		APIKeyAuthenticator: auth.CreateAuthenticator(Options.Auth.EnableAuth),
```
These should come from the auth instance that will be initialized from the above comment

ronniel1 commented on deploy/bm-inventory-configmap.yaml:
```diff
@@ -9,4 +9,6 @@ data:
-  BASE_DNS_DOMAINS: REPLACE_DOMAINS # example: name1:id1/provider1,name2:id2/provider2
\ No newline at end of file
+  BASE_DNS_DOMAINS: REPLACE_DOMAINS # example: name1:id1/provider1,name2:id2/provider2
+  ENABLE_AUTH: IS_AUTH_ENABLED
```
IS_AUTH_ENABLED -> REPLACE_AUTH_ENABLED_FLAG

nmagnezi replied:
Done

ronniel1 commented on pkg/auth/auth_handler.go:
```diff
@@ -0,0 +1,176 @@
+	LastName  string `json:"last_name"`
+	Email     string `json:"email"`
+	Issuer    string `json:"iss"`
+	ClientID  string `json:"clientId"`
```
Should have org as well

nmagnezi replied:
Done

ronniel1 commented on pkg/auth/auth_handler.go:
```diff
@@ -0,0 +1,176 @@
+	authHandler.CertURL = certURL
+	err := authHandler.populateKeyMap()
+	if err != nil {
+		logrus.Fatalln("Failed to init auth manager,", err)
```
All logrus usage should change to logger passed during initialization

nmagnezi replied:
Done

ronniel1 commented on pkg/auth/helpers.go:
```diff
@@ -0,0 +1,133 @@
+			},
+		},
+	}
+	glog.V(5).Infof("Getting JWK public key from %s", url)
```
use logrus

nmagnezi replied:
Done

ronniel1 commented on pkg/auth/auth_handler.go:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	// Try to read the JWT public key object file.
+	a.keyMap, err = downloadPublicKeys(a.CertURL, trustedCAs)
```
Just a suggestion.
I  think that you will find this hard to UT if you don't mock the downloadPublicKeys()
Maybe it would help it the auth_utils was an interface that supplied the downloadPublicKeys.
and AuthHandler uses the interface. Than you will be able to mock it in the UTs

nmagnezi replied:
I followed your advice here, let me know if it looks ok to you before I click `resolve` on this one.

ronniel1 commented on pkg/auth/auth_handler.go:
```diff
@@ -0,0 +1,176 @@
+
+func (a *AuthHandler) getValidationToken(token *jwt.Token) (interface{}, error) {
+	// Try to get the token kid.
+	kid, ok := token.Header["kid"]
```
What is 'kid'?

nmagnezi replied:
`kid` field will tell you which signing key should be used for verification. ([Source](https://www.jsonwebtoken.io/))

ronniel1 commented on pkg/auth/auth_handler.go:
```diff
@@ -0,0 +1,176 @@
+		return nil, fmt.Errorf("no kid found in jwt token")
+	}
+
+	// Try to get currect cert from certs map.
```
typo currect -> correct

nmagnezi replied:
Done

ronniel1 commented on swagger.yaml:
```diff
@@ -226,6 +252,8 @@ paths:
       tags:
         - installer
+      security:
+        - userAuth: []
```
This will require both userAuth and agentAuth

@avishayt  Am I right?

Need to verify that we can specify both, and it will work when any 1 of the two is supplied

ronniel1 commented on swagger.yaml:
```diff
@@ -806,6 +866,8 @@ paths:
       tags:
         - installer
+      security:
+        - userAuth: [] # TODO: confirm this
```
userAuth

ronniel1 commented on swagger.yaml:
```diff
@@ -848,6 +910,8 @@ paths:
       tags:
         - managed_domains
+      security:
+        - userAuth: [] # TODO: confirm this
```
userAuth

ronniel1 commented on swagger.yaml:
```diff
@@ -262,6 +290,8 @@ paths:
     get:
```
Since userAuth is used for most APIs it can be set as a default auth, no need to set it at each API

nmagnezi replied:
Done

ronniel1 reviewed (commented):
Nice work.
Need extensive UTs

Commit:
MGMT-773 Implement Swagger APIKeyAuthenticator Based Auth Handler

Commit:
MGMT-733 Adds Swagger API Key Security

Bumped Swagger image version to 0.25.0
Bumped go-openapi/runtime to v0.19.20 to match the Swagger image version.

Commit:
MGMT-733 Swagger Generate Server and Client

nmagnezi reviewed (commented):
Addresses multiple comments (and will shortly address the rest).
Unit tests are not included as they are still WIP.

ronniel1 commented on cmd/main.go:
```diff
@@ -115,7 +117,7 @@ func main() {
 		log.Fatal("failed to auto migrate, ", err)
 	}
-
+	authHandler := auth.NewAuthHandler(Options.Auth, (log.WithField("pkg", "auth")))
```
No need for the parenthesis around the log arg

nmagnezi replied:
Done

Commit:
MGMT-733 Auth Handler Unit Tests

Commit:
MGMT-773 Adapt subsystem and unit tests to use Auth

nmagnezi commented on subsystem/subsystem_suite_test.go:
```diff
@@ -21,25 +23,39 @@ var log *logrus.Logger
+	agentKey := "X-Secret-Key"
+	agentKeyValue := "SecretKey"
+	//userKey := "Authorization"
+	//userKeyValue := "userKey"
```
to be removed. I'm testing this

nmagnezi commented on tools/deploy_assisted_installer.py:
```diff
@@ -7,10 +7,21 @@
+        with open(KEY_FILE, "r") as f:
+            return f.read()
+    except Exception as e:
+        print("Got exception {}, when tried to read Could not read key file at {}."
```
will remove `Could not read`
